accepted/2022/enable-512-vectors.md

```diff
@@ -0,0 +1,518 @@
+
+| Method  |
+| ------ | 
+| `VectorMask<T> Vector<T>.Add(Vector<T> v1, Vector<T> v2, VectorMask<T> cond)` | 
+| `VectorMask<T> Vector<T>.And(Vector<T> v1, Vector<T> v2, VectorMask<T> cond)` | 
+
+(OPEN: What API methods to extend here?)
+
+
+
+#### MaskExpr API
+
+`MaskExpr` is meant to start a small domain specific language (DSL) for Vector SIMD processing. To keep the implementation simple, we choose to directly use a `C#` expression (embedded DSL) encoded through a lambda. 
+
+In the following example:
+
+```C#
+public Vector256<int> SumVector256(ReadOnlySpan<int> source)
+{
+  Vector256<int> vresult = Vector256<int>.Zero;
+
+  for (int i = 0; i < source.Length;; i += Vector256<int>.Count)
+  {
+    Vector256<int> v1 = new Vector256<int>(source.Slice(i));
+    VectorMask256<int> mask = v1.MaskExpr(x => x > 0 & x != 5);
+    vresult = Vector256.Add(vresult, v1, mask);
+  }
+
+  return vresult;
+}
+```
+
+to JIT should logically create a `VectorMask` equivalent to `v1.GreaterThanByVectorMask(0) & v1.NotEqualByVectorMask(5)`.
+
+Given a usage of `MaskExpr`, `v.MaskExpr(x => e)`, where `v` represents a `Vector` and `e` represents an expression:
+
+- `v` defines the vector that we perform the conditional SIMD for (the implementation determines how the masking must be done)
+
+- `x` can be logically thought of as `v[i]` and is used to hint to the JIT on how to build the mask.
+
+- `e` defines an expression that the JIT will lift up in an appropriate masking. `e` must be a boolean expression, and is restricted to the standard boolean comparision operations.
+
+- If `e` does not type check to a boolean expression, and uses expressions other than boolean comparison operations and `x`, the JIT will throw `OperationNotSupportedError`.
+
+There are more advanced uses of a true "embedded DSL" and if the idea seems applicable, we can extend `MaskExpr` to handle many more interesting cases (at the cost of more advanced lifting from C# expression to something the JIT can work with).
+
+#### Leading and Trailing Processing API Methods
+
+We propose the following methods for each `Vector` API for processing leading and trailing elements. 
+
+| Method  |
+| ------ | 
+| `VectorMask<T> Vector<T>.CreateLeadingMask(int rem)` | 
+| `VectorMask<T> Vector<T>.LoadLeading(Span<T> v1, VectorMask<T> mask)` | 
+| `VectorMask<T> Vector<T>.StoreLeading(Span<T> v1, VectorMask<T> mask)` | 
+| `VectorMask<T> Vector<T>.CreateTrailingMask(int rem)` | 
+| `VectorMask<T> Vector<T>.LoadTrailing(Span<T> v1, VectorMask<T> mask)` | 
+| `VectorMask<T> Vector<T>.StoreTrailing(Span<T> v1, VectorMask<T> mask)` | 
+
+### Internals Upgrades for EVEX/AVX512 Enabling
+
+Broadly speaking, we can break the implementation of 512-bit vectors and `VectorMask` into the following components:
+
+1. Enable EVEX encoding for Vector128/Vector256 in the xarch emitter.
```

Comment:
We should track a `0. Enable VM recognition and tracking of the AVX-512 ISAs`

This basically requires updating `vm/codeman.cpp` to check the relevant CPUID bits and set the appropriate `InstructionSetFlags`. It will also require updating CG2 to track the relevant bits which will require extending the existing flags support beyond a single `int64_t` field.

Reply:
For 1. It might be good to clarify this does not include `kmask` support.

Reply:
Am I correct in understanding that this step is mostly about adding the AVX512 specific AVX512VL instructions (lets say, `vpermb`) with the rest of the EVEX encoding work coming later? Or do you think we would template out a lot of the EVEX encoding here, i.e., ability to encode k0-k7 in the EVEX opcode, with the actual support for k0-k7 coming in a later step?

To clarify, what I am asking is, do we plan to do the bulk of the EVEX opcode work in this stage, or do we plan to do the minimal to get some AVX512VL instructions up and running, and continue to flesh out EVEX opcode encoding as we add the rest of the EVEX features later on.

Reply:
I think breaking it apart makes sense, but we need the `0.` step as the first bit. Without it we don't have a way for the JIT to determine whether or not emitting the instructions are legal in the first place nor the corresponding tracking of their usage for Ready To Run/Crossgen scenarios.